### ISSUE TITLE
Trigger optogenetics through USB-6002 NIDaq

### ISSUE THREAD

**Issue by XX-Yin** [AllenNeuralDynamics/dynamic-foraging-task#11]
@bruno-f-cruz @jsiegle 

Hi Bruno,

Previously, we can set the trigger source of PCIe-6323 DAQ using the code you provided (shown below; related to [discussion](https://github.com/AllenNeuralDynamics/BonsaiBehavior/discussions/77) ). 

1) Can we do the same thing for USB DAQ-6002? The USB DAQ-6002 have ports like shown in the second image, and can be accessed in bonsai through DAQmx node.  

2) How to select the analog output port (e.g. AO0 or AO1) in the customized node "TriggerAnalogOutput"?

3) Can we simultaneously load different buffers to AO0 and AO1, and set AO0 and AO1 to be triggered by the same port? We need to do optogenetics for two locations in some cases, which means we need 2 waveforms. 

![image](https://github.com/AllenNeuralDynamics/dynamic-foraging-task/assets/109394934/8e1e478b-747c-4ea9-986f-304f797b7d0a)

![image](https://github.com/AllenNeuralDynamics/dynamic-foraging-task/assets/109394934/4366b0d4-f9d7-45bf-b726-72024cd948f8)

**Comment by bruno-f-cruz** [comment]
Hi,

1) No idea. Bonsai is agnostic as to whether the task is possible to run or not. Once you try run it for the first time, there will be an API call to "validate" the task. If the board is not able to perform it due to hardware constraints it will throw an error. So I guess, just give it a try!

2) You access the AnalogOutput pins the same way you would with the other PCI board. Just go to "Channels" and select your IO.

3) Not sure, I would have to spend more time trying to do this but it seems easy enough. Instead of selecting a single channel, select two and send a matrix with two rows to the operator. As long as the trigger for both channels (ie. the trigger for the task) can be the same I think it should work.

**Comment by XX-Yin** [comment]
Hi Bruno,

1. It works for the USB daq, but the channel name seems to be different. "PFI0" in PCIe-6323 is corresponding to P2.0 in USB DAQ-6002, and "PFI1" in PCIe-6323 is corresponding to P1.1 in USB DAQ-6002. 

2. It also works. The same as other PCI board. 

3. For this part, can you give me an example workflow to demonstrate it?

**Comment by XX-Yin** [comment]
I had another issue to repeatedly trigger "TriggeredAnalogOutput". The workflow is shown below. It works when I first trigger the workflow, and I can get the analogoutput from NIdaq. However, the second time the chain is triggered, it immediately crashes (the error information is shown below). It seems that if I hope to save buffer to TriggeredAnalogOutput the second time, I must restart the bonsai workflow.

![image](https://github.com/AllenNeuralDynamics/dynamic-foraging-task/assets/109394934/72ee098e-6e9c-4c33-aca8-0e49bfa1efa3)

**Comment by bruno-f-cruz** [comment]
> 3. For this part, can you give me an example workflow to demonstrate it? For example, how to concatenate Waveform1_1 and Waveform1_2 and send them to the same TriggerAnalogOutput node?

You will need to define the channels as I don't have a nidaq board with me.
![image](https://github.com/AllenNeuralDynamics/dynamic-foraging-task/assets/7049351/43a7218c-9f6d-4a63-a689-b6cebe5cf23f)


<details>
    <summary>Workflow (Copy+Paste me)</summary>
    
```xml



<?xml version="1.0" encoding="utf-8"?>
<WorkflowBuilder Version="2.7.3"
                 xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                 xmlns:dsp="clr-namespace:Bonsai.Dsp;assembly=Bonsai.Dsp"
                 xmlns:mx="clr-namespace:Bonsai.DAQmx;assembly=Bonsai.DAQmx"
                 xmlns:rx="clr-namespace:Bonsai.Reactive;assembly=Bonsai.Core"
                 xmlns="https://bonsai-rx.org/2018/workflow">
  <Workflow>
    <Nodes>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="dsp:FunctionGenerator">
          <dsp:BufferLength>3000</dsp:BufferLength>
          <dsp:Frequency>100</dsp:Frequency>
          <dsp:Waveform>Sine</dsp:Waveform>
          <dsp:SampleRate>44100</dsp:SampleRate>
          <dsp:Amplitude>1</dsp:Amplitude>
          <dsp:Offset>0</dsp:Offset>
          <dsp:Phase>0</dsp:Phase>
        </Combinator>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="dsp:FunctionGenerator">
          <dsp:BufferLength>3000</dsp:BufferLength>
          <dsp:Frequency>100</dsp:Frequency>
          <dsp:Waveform>Sine</dsp:Waveform>
          <dsp:SampleRate>30000</dsp:SampleRate>
          <dsp:Amplitude>1</dsp:Amplitude>
          <dsp:Offset>0</dsp:Offset>
          <dsp:Phase>0</dsp:Phase>
        </Combinator>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="rx:Zip" />
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="dsp:Concat">
          <dsp:Axis>0</dsp:Axis>
        </Combinator>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="mx:AnalogOutput">
          <mx:SignalSource />
          <mx:SampleRate>0</mx:SampleRate>
          <mx:ActiveEdge>Rising</mx:ActiveEdge>
          <mx:SampleMode>ContinuousSamples</mx:SampleMode>
          <mx:BufferSize>1000</mx:BufferSize>
          <mx:Channels>
            <mx:AnalogOutputChannelConfiguration>
              <mx:ChannelName />
              <mx:MinimumValue>-10</mx:MinimumValue>
              <mx:MaximumValue>10</mx:MaximumValue>
              <mx:VoltageUnits>Volts</mx:VoltageUnits>
            </mx:AnalogOutputChannelConfiguration>
            <mx:AnalogOutputChannelConfiguration>
              <mx:ChannelName />
              <mx:MinimumValue>-10</mx:MinimumValue>
              <mx:MaximumValue>10</mx:MaximumValue>
              <mx:VoltageUnits>Volts</mx:VoltageUnits>
            </mx:AnalogOutputChannelConfiguration>
          </mx:Channels>
        </Combinator>
      </Expression>
    </Nodes>
    <Edges>
      <Edge From="0" To="1" Label="Source1" />
      <Edge From="0" To="2" Label="Source2" />
      <Edge From="1" To="2" Label="Source1" />
      <Edge From="2" To="3" Label="Source1" />
      <Edge From="3" To="4" Label="Source1" />
    </Edges>
  </Workflow>
</WorkflowBuilder>


```

</details>

**Comment by bruno-f-cruz** [comment]
> I had another issue to repeatedly trigger "TriggeredAnalogOutput". The workflow is shown below. It works when I first trigger the workflow, and I can get the analogoutput from NIdaq. However, the second time the chain is triggered, it immediately crashes (the error information is shown below). It seems that if I hope to save buffer to TriggeredAnalogOutput the second time, I must restart the bonsai workflow.
> 
> ![image](https://user-images.githubusercontent.com/109394934/241086373-72ee098e-6e9c-4c33-aca8-0e49bfa1efa3.png)

I think that there is an option in the API to recycle the trigger. Can you send me python or matlab code of how you would set this up? I am not familiar with the features of the board so it would be useful to have something as reference. thanks

**Comment by XX-Yin** [comment]
> > 3. For this part, can you give me an example workflow to demonstrate it? For example, how to concatenate Waveform1_1 and Waveform1_2 and send them to the same TriggerAnalogOutput node?
> 
> You will need to define the channels as I don't have a nidaq board with me. ![image](https://user-images.githubusercontent.com/7049351/241139974-43a7218c-9f6d-4a63-a689-b6cebe5cf23f.png)
> 
> Workflow (Copy+Paste me)

1) It only works for "ContinousSamples", not for "FiniteSamples". 
2) When I tried to zip WaveForm1_2 and WaveForm1_1 and sent to "TriggeredAnalogOutput", it showed an error (shown below).

![image](https://github.com/AllenNeuralDynamics/dynamic-foraging-task/assets/109394934/00260667-5f4f-4a5e-997b-ad39417963a1)

**Comment by bruno-f-cruz** [comment]
1. Not sure what this refers to?
2. The output of zip is a Tuple not a matrix, you need to concatenate the two matrices with "Concat", like I showed you in the previous example.

**Comment by XX-Yin** [comment]
> thanks



> 

Hi Bruno, 

It's the [dynamic foraging task](https://github.com/AllenNeuralDynamics/dynamic-foraging-task/). The main change should be in the group "Optogenetics2". 

The final purpose is to trigger preloaded buffer in NI daq for optogenetics. There are 2 issues remain to solve. 

1) simultaneously trigger 2 NIdaq channels. 
2) the "TriggeredAnalogOutput" can be triggered repeatedly.

**Comment by bruno-f-cruz** [comment]
haha, sorry I meant I didnt know what:
> "It only works for "ContinousSamples", not for "FiniteSamples"."
referred to

**Comment by XX-Yin** [comment]
> haha, sorry I meant I didnt know what:
> 
> > "It only works for "ContinousSamples", not for "FiniteSamples"."
> 
> referred to

If I select "SampleMode" in the "TriggeredAnalogOutput" node as  "ContinousSamples", it will show an error as shown below:
 
![image](https://github.com/AllenNeuralDynamics/dynamic-foraging-task/assets/109394934/181213a0-fa20-424e-8b41-e9e27b75007b)

**Comment by bruno-f-cruz** [comment]
Have you tried to do what the error suggested and double your buffer size
to 2000?

**Comment by XX-Yin** [comment]
Still the same error information.

**Comment by bruno-f-cruz** [comment]
What are the size of the matrices you are trying to playback? Also, can you send me a python or matlab snippet of how you would set this up there? It would really help!

**Comment by XX-Yin** [comment]
I solved the issue "It only works for "ContinousSamples", not for "FiniteSamples"." by adding repeat and take nodes after function generator. 

![image](https://github.com/AllenNeuralDynamics/dynamic-foraging-task/assets/109394934/40a99a6e-ce3b-41b6-a77d-ccbf54dbf863)

**Comment by bruno-f-cruz** [comment]
Ok, you are likely not matching the size of the matrix with the size of the buffer the nidaq expects. Repeat keeps generating buffers until you fill the expected size. It will "work" in the sense that the task will play, but you will likely not output the waveform you expect.

**Comment by XX-Yin** [comment]
Thanks for explanation. I also tried to modify the dynamic foraging workflow. Currently, I can simultaneously get two NIdaq outputs. The remaining issue is to repeatedly trigger "TriggeredAnalogOutput" .

![image](https://github.com/AllenNeuralDynamics/dynamic-foraging-task/assets/109394934/8fb5e59b-20bb-458c-bde4-fdcd2d02650b)

**Comment by bruno-f-cruz** [comment]
Again, can you send me a snippet of how this would look like in python or matlab? I don't have a nidaq with me, so it would be hard to iterate, if you already have a working example from another language it will be much easier!

**Comment by XX-Yin** [comment]
You mean the communication between python and Bonsai (how to trigger the bonsai workflow in python)?

**Comment by bruno-f-cruz** [comment]
Nop, I thought you had some nidaq tasks working on matlab and python, but I might be wrong. If that is not the case this might require taking some time to figure out how to construct the task you are interested in.

**Comment by XX-Yin** [comment]
Yes, we have a task. It's the dynamic foraging task.  https://github.com/AllenNeuralDynamics/dynamic-foraging-task/. 

Is this what you mean?

**Comment by bruno-f-cruz** [comment]
Sorry for the confusion, but what I meant is whether you have an API alternative to interact with the nidaq boards. I seem to remember that a few months ago some of you were using the python API to interact with it (I might be mistaken sorry!)
e.g. https://nidaqmx-python.readthedocs.io/en/latest/

**Comment by XX-Yin** [comment]
For the current workflow, we don’t use the API. I think Josh used it in a different workflow for a different NIdaq. How will it help you to solve the issue? You mean we can directly use the python API to control the NIdaq instead of using bonsai?

**Comment by bruno-f-cruz** [comment]
The settings from the python API can be easily translated into the c# one. If you know the code to achieve what you need in python I might be able to translate it into c# and make it work with bonsai.

**Comment by XX-Yin** [comment]
I currently have no idea how to achieve this in python. I can actually get correct analog outputs from NIdaq using workflow below, but it can be triggered only once. Do you have any ideas about the reason? 
![](https://user-images.githubusercontent.com/109394934/241324510-8fb5e59b-20bb-458c-bde4-fdcd2d02650b.png)

**Comment by XX-Yin** [comment]
Thanks! Have a great weekend!

**Comment by XX-Yin** [comment]
> Thanks! Have a great weekend!



> I suspect there is a modification parameter in the nidaq task that makes it possible to retrigger. If I have a chance I will try to look at it next week.
> […](#)

Just to clarify it. The issue is not to trigger the same buffer second time, and I will send a new buffer to Nidaq for each new trigger. The error happened when I sent the new buffer to NIdaq after finish of the first trigger.

**Comment by bruno-f-cruz** [comment]
What happens when you match the buffer sizes?

**Comment by XX-Yin** [comment]
I think the buffer size is matched as it works for the first trigger.And I don’t change the waveform length for the second buffer.

**Comment by bruno-f-cruz** [comment]
With or without the repeat node? Can you share that portion of the code?
Thanks

**Comment by XX-Yin** [comment]
It’s in the group Optogenetics2 of this dynamic foraging task. I already pushed it to GitHub.

**Comment by bruno-f-cruz** [comment]
Can you try the following snippet?
If it does not work, can you try to make the `BufferSize` property in the `TriggeredAnalogOutput` equal 10000? 

<details>
    <summary>Workflow (Copy+Paste me)</summary>
    
```xml

<?xml version="1.0" encoding="utf-8"?>
<WorkflowBuilder Version="2.7.3"
                 xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                 xmlns:p1="clr-namespace:Bonsai.DAQmx;assembly=Extensions"
                 xmlns:dsp="clr-namespace:Bonsai.Dsp;assembly=Bonsai.Dsp"
                 xmlns:rx="clr-namespace:Bonsai.Reactive;assembly=Bonsai.Core"
                 xmlns="https://bonsai-rx.org/2018/workflow">
  <Workflow>
    <Nodes>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="dsp:ScalarBuffer">
          <dsp:Size>
            <dsp:Width>5000</dsp:Width>
            <dsp:Height>1</dsp:Height>
          </dsp:Size>
          <dsp:Depth>F32</dsp:Depth>
          <dsp:Channels>1</dsp:Channels>
          <dsp:Value>
            <dsp:Val0>2.5</dsp:Val0>
            <dsp:Val1>0</dsp:Val1>
            <dsp:Val2>0</dsp:Val2>
            <dsp:Val3>0</dsp:Val3>
          </dsp:Value>
        </Combinator>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="dsp:ScalarBuffer">
          <dsp:Size>
            <dsp:Width>5000</dsp:Width>
            <dsp:Height>1</dsp:Height>
          </dsp:Size>
          <dsp:Depth>F32</dsp:Depth>
          <dsp:Channels>1</dsp:Channels>
          <dsp:Value>
            <dsp:Val0>5</dsp:Val0>
            <dsp:Val1>0</dsp:Val1>
            <dsp:Val2>0</dsp:Val2>
            <dsp:Val3>0</dsp:Val3>
          </dsp:Value>
        </Combinator>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="rx:Zip" />
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="dsp:Concat">
          <dsp:Axis>0</dsp:Axis>
        </Combinator>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="p1:TriggeredAnalogOutput">
          <p1:SignalSource />
          <p1:TriggerSource>/Dev1/PFI0</p1:TriggerSource>
          <p1:SampleRate>5000</p1:SampleRate>
          <p1:ActiveEdge>Rising</p1:ActiveEdge>
          <p1:SampleMode>FiniteSamples</p1:SampleMode>
          <p1:BufferSize>5000</p1:BufferSize>
          <p1:Channels>
            <p1:AnalogOutputChannelConfiguration>
              <p1:ChannelName />
              <p1:MinimumValue>-10</p1:MinimumValue>
              <p1:MaximumValue>10</p1:MaximumValue>
              <p1:PhysicalChannel>Dev1/ao0</p1:PhysicalChannel>
              <p1:VoltageUnits>Volts</p1:VoltageUnits>
            </p1:AnalogOutputChannelConfiguration>
            <p1:AnalogOutputChannelConfiguration>
              <p1:ChannelName />
              <p1:MinimumValue>-10</p1:MinimumValue>
              <p1:MaximumValue>10</p1:MaximumValue>
              <p1:PhysicalChannel>Dev1/ao1</p1:PhysicalChannel>
              <p1:VoltageUnits>Volts</p1:VoltageUnits>
            </p1:AnalogOutputChannelConfiguration>
          </p1:Channels>
        </Combinator>
      </Expression>
    </Nodes>
    <Edges>
      <Edge From="0" To="2" Label="Source2" />
      <Edge From="1" To="2" Label="Source1" />
      <Edge From="2" To="3" Label="Source1" />
      <Edge From="3" To="4" Label="Source1" />
    </Edges>
  </Workflow>
</WorkflowBuilder>


```

</details>

**Comment by XX-Yin** [comment]
It works for the following workflow. 

![image](https://github.com/AllenNeuralDynamics/dynamic-foraging-task/assets/109394934/ba2eab83-bf98-471e-8fa4-f781c082e1c1)

However, how is it applied to the workflow below? 
![image](https://github.com/AllenNeuralDynamics/dynamic-foraging-task/assets/109394934/128284ee-8b05-4d90-8eab-9d068020e978)

**Comment by bruno-f-cruz** [comment]
As I mentioned before, set the buffer size of the AnalogOutput operator to be the same as the incoming buffer size.

**Comment by XX-Yin** [comment]
It's not the problem of the buffer size. I am pretty sure the buffer size is correct. The problem is that the second time (trial) I send Waveform1_2 and WaveForm1_1 to TriggeredAnalogOuput, it has the error shown below.

![image](https://github.com/AllenNeuralDynamics/dynamic-foraging-task/assets/109394934/6dbbd395-92bd-4a88-bf79-4a8ce245d9d6)

**Comment by XX-Yin** [comment]
I think I narrowed down the issue. 

If I set the buffer size very large (much larger than the input waveform size), the TriggeredAnalogOuput can be triggered multiple times, but the workflow can't be stopped until the buffer size is used up. So the issue is probably related to the buffer size can't be reset each trial through "Location1Size".

**Comment by bruno-f-cruz** [comment]
Ok, I am a bit confused... What do you mean the buffer size cant be reset? When you set a task in "FiniteSamples" mode, AFAIK, the buffer will not get replaced as you send more data. If you have a buffer of 5000, and you send a waveform of 5000 points, if you send an extra sample it will crash the task. That's the whole point of using "FiniteSamples" no? If you want to use a different waveform each time, should you not just create a new task on each update of the buffer?

**Comment by XX-Yin** [comment]
So why doesn't the buffer reset every trial by Location1Size ?

![image](https://github.com/AllenNeuralDynamics/dynamic-foraging-task/assets/109394934/2f7b2bdb-fd87-4c51-9304-21d23fd4d1af)

**Comment by XX-Yin** [comment]
Currently, the waveform length in each trial could be different, so the strategy is to reset the buffer size of finite buffers each trial through Location1Size. 

I think that I may find the reason of this error. It seems that Buffersize determined by Location1Size was reset after WaveForm1_2 and WaveForm1_1, so there was no buffer size set when I tried to set the buffer size through WaveForm1_1 and WaveForm1_2. This may be solved by group TriggeredAnalogOutput using selectmany.

**Comment by bruno-f-cruz** [comment]
Indeed, if you want to reset the nidaq task on every single trial, you must
close it after you are done with it and restart it. Probably inside a
SelectMany.

**Comment by XX-Yin** [comment]
It proved my hypothesis. It works now! I think the error is related to the update order between Buffersize and Waveform  (also related to restart it through selectmany), and difficult to locate. Thank you Bruno! 

![image](https://github.com/AllenNeuralDynamics/dynamic-foraging-task/assets/109394934/59748ede-cddd-457d-973f-1c74c507f9a0)

**Comment by XX-Yin** [comment]
Solved now!
